### PR TITLE
roachtest: redirect CRDB logs to roachprod test logger

### PR DESCRIPTION
**Pull request by herkolategan**
Previously, the CRDB logger was configured to log to a file to catch any logs
written to it during a roachtest run. This has two issues: It is not easy to
discern under which test a log entry was made, and does not follow the regular
structure of roachtest logs; sensitive information is also possibly leaked due
to roachtest program arguments being written by the file sink based CRDB logger.

One possible approach could be to disallow the CRDB logger in roachtests. But
there are a few usages through transitive dependencies, such as `pkg/testutils`
that write to CRDB log. There are also cases where external libraries use the
standard log which CRDB log will intercept. Considering all these scenarios this
PR presents an alternative to intercept CRDB logs and use a best effort approach
to log it under the correct test logger, or use a fallback log if no such logger
could be determined.

To bridge the CRDB log to a test logger we use the goroutine stacks. A bridge
can be set up for a goroutine (ID), and a best effort approach is taken to catch
logs from child goroutines. This feature is only supported from go 1.21, which
added the goroutine origin to the `created by` section of a stack. But it can
also be derived by enabling GODEBUG="tracebackancestors=N".

Informs: #131412

Epic: None
Release note: None

### CODE REVIEW

**Comment by herkolategan** [comment]
TC run 0.2: https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_RoachtestNightlyGceBazel/17101844:id/cdc/tpcc-1000/sink%3Dkafka/run_1/artifacts.zip!/test.log
Kafka log redirected to the appropriate test.

**Comment by herkolategan** [comment]
> Do those usages depend on some (crdb log) feature which isn't available in the roachtest log?

No, it doesn't use a specific feature. But the CRDB is log is accessed 'statically' through some of the `testutilts` in the root pkg (not testutils in `roachtest). 

> By replacing crdb log with roachtest, the standard log would then be redirected to roachtest, right? I guess I'm wondering why you didn't choose this approach? Is it much more involved to replace crdb log and forbid its use inside roachtest?

I'm starting to think the approach I took may be overkill, as it primarily catches kafka's 3rd party logs, and succeedssoon / retry from the cockroach testutils. The alternate would be to rather have an implementation of those two utilities under roachtest that takes a roachprod/roachtest logger scoped to the test. And have our own bridge for the standard log in roachtest to prevent/redirect the noisy logs from Kafka. We can then prevent the import of "util/log" with bazel, but it still won't stop anyone from using a cockroach testutil that could end up calling cockroach logs, but I suppose we can catch those in code reviews. Somehow I feel it might keep on happening as it's easy to reach for something in testutils, and since the unit tests use it, it's normal to use CRDB log in those utils. Might need to ban "pkg/testutils" import as well?

The log bridge in this solution works, but I don't like using the goroutine IDs to scope the log bridges, unfortunately this part is a bit hacky.

TL;DR; definitely not attached to this solution, could go the route of eliminating / replacing the use of 2 small `testutils` from cockroach, and then denying "util/log" import.

**Comment by herkolategan** [comment]
Alternatively, the original problem this was trying to solve is to not use the file sink log on CRDB as it leaks program arguments. We could also keep the existing solution and opt to pass sensitive arguments as env vars?

**Comment by herkolategan** [comment]
Thought about this some more. We can ban the CRDB log import on `roachtest`, and move this PR's log redirect implementation next to the roachrpod logger that it is actually bridging. Since we have more control over `roachprod`, the likelihood of a stray CRDB log introduced there is much less than in `roachtest`.

**Comment by herkolategan** [comment]
Closing this PR, as I plan on restructuring it a bit, and removing the goid parts, to opt for a simpler dump to file.
